File: tests/chaos/next_use_semantic_preflight.c
/* Test-only fixed value wire. Journal bytes are never code. NGPL.
 * Include the unmodified production implementation in this test TU solely for
 * read-only replay-state inspection: exported snapshots describe LIVE state,
 * not the staged validator. No new production setter or replay-state writes.
 */
#include "hack.h"
#include "chaos_next_use_runtime.h"
#include <stdint.h>
#include <limits.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
long moves, monstermoves;
boolean panicking;
/* Same minimal I/O/panic stubs as next_use_replay.c. */
void panic(const char *s, ...) { (void)s; abort(); }
void bwrite(int fd, genericptr_t p, unsigned n) {
    if (write(fd, p, n) != (ssize_t)n) abort();
}
int chaos_next_use_mread(int fd, void *p, unsigned n) {
    return read(fd, p, n) == (ssize_t)n;
}
void mread(int fd, genericptr_t p, unsigned n) {
    if (!chaos_next_use_mread(fd, p, n)) abort();
}
#include "../../src/chaos_next_use_runtime.c"

static int number(FILE *in, uint64_t *v, uint64_t max) {
    unsigned char b[8]; int i;
    if (fread(b, 1, 8, in) != 8) return 0;
    *v = 0;
    for (i = 0; i < 8; ++i) *v = (*v << 8) | b[i];
    return *v <= max;
}
static int text(FILE *in, char *s, size_t cap) {
    uint64_t n;
    if (!number(in, &n, cap - 1) || fread(s, 1, n, in) != n) return 0;
    if (memchr(s, 0, n)) return 0;
    s[n] = 0; return 1;
}
#define GET(field, max) do { if (!number(in, &v, max)) return 0; field = v; } while (0)
static int record_read(FILE *in, struct chaos_next_use_replay_input *r) {
    uint64_t v; int i;
    memset(r, 0, sizeof *r);
    GET(r->replay_input_v, INT_MAX);
    GET(r->expected_last_root, LONG_MAX);
    GET(r->activation_move, LONG_MAX);
    GET(r->notice_root, LONG_MAX);
    GET(r->witness_notice_seq, LONG_MAX);
    GET(r->token_present, INT_MAX);
    GET(r->root_present, INT_MAX);
    GET(r->expected_result, INT_MAX);
    GET(r->published, INT_MAX);
    GET(r->pre_public, INT_MAX);
    GET(r->operation, INT_MAX);
    GET(r->family, INT_MAX);
    GET(r->root, LONG_MAX);
    if (!text(in, r->source_sha256, sizeof r->source_sha256)) return 0;
    GET(r->callback_ordinal, INT_MAX);
    GET(r->state, INT_MAX);
    GET(r->seq, INT_MAX);
    GET(r->slot_w, INT_MAX);
    GET(r->slot_f, INT_MAX);
    GET(r->w_runtime, INT_MAX);
    GET(r->m_id, UINT_MAX);
    GET(r->at_move, LONG_MAX);
    GET(r->fountain_outcome, INT_MAX);
    GET(r->run_token, LONG_MAX);
    GET(r->level_token, LONG_MAX);
    GET(r->origin_w_live, INT_MAX);
    GET(r->origin_f_live, INT_MAX);
    GET(r->whistle_count, INT_MAX);
    GET(r->fountain_count, INT_MAX);
    GET(r->end_reason, INT_MAX);
    GET(r->expected_attention, INT_MAX);
    GET(r->decision_root, LONG_MAX);
    GET(r->cursor, 4096);
    GET(r->manifest_root, LONG_MAX);
    GET(r->notice_seq, LONG_MAX);
    GET(r->end_seq, LONG_MAX);
    GET(r->manifestation_delivered, INT_MAX);
    GET(r->displaced, INT_MAX);
    GET(r->invalid, INT_MAX);
    GET(r->private_count, INT_MAX);
    GET(r->public_count, INT_MAX);
    GET(r->post.phase, INT_MAX);
    GET(r->post.delay_used, INT_MAX);
    GET(r->post.delay_until, INT_MAX);
    GET(r->post.termination_emitted, INT_MAX);
    GET(r->post.identity_unsafe, INT_MAX);
    GET(r->post.pending_w_capture, INT_MAX);
    GET(r->post.f_inflight, INT_MAX);
    GET(r->post.witnessed, INT_MAX);
    GET(r->post.attention_claimed, INT_MAX);
    GET(r->post.callback_w, INT_MAX);
    GET(r->post.callback_f, INT_MAX);
    GET(r->post.whistle_count, INT_MAX);
    GET(r->post.fountain_count, INT_MAX);
    GET(r->post.origin_w_live, INT_MAX);
    GET(r->post.origin_f_live, INT_MAX);
    GET(r->post.manifest_success, INT_MAX);
    GET(r->post.armed_m_id, UINT_MAX);
    GET(r->post.expected_manifest_m_id, UINT_MAX);
    GET(r->post.activation_monstermoves, LONG_MAX);
    GET(r->post.armed_root, LONG_MAX);
    GET(r->post.pending_w_root, LONG_MAX);
    GET(r->post.f_root, LONG_MAX);
    GET(r->post.current_run_token, LONG_MAX);
    GET(r->post.current_level_token, LONG_MAX);
    GET(r->post.expected_manifest_root, LONG_MAX);
    GET(r->post.expected_notice_seq, LONG_MAX);
    GET(r->post.expected_end_seq, LONG_MAX);
    GET(r->expected_token.root, LONG_MAX);
    GET(r->expected_token.active, INT_MAX);
    GET(r->expected_token.remap, INT_MAX);
    GET(r->expected_token.consumed, INT_MAX);
    if (r->private_count > 4 || r->public_count > 1) return 0;
    for (i = 0; i < r->private_count; ++i) {
        struct chaos_next_use_runtime_private_record *p = &r->private_records[i];
    GET(p->next_use_private_v, INT_MAX);
    GET(p->kind, INT_MAX);
    GET(p->seq, INT_MAX);
    GET(p->at_move, INT_MAX);
    GET(p->program_id, INT_MAX);
    if (!text(in, p->source_sha256, sizeof p->source_sha256)) return 0;
        switch (p->kind) {
        case 3:
    GET(p->data.intent.callback_ordinal, INT_MAX);
    GET(p->data.intent.trigger, INT_MAX);
    GET(p->data.intent.validation, INT_MAX);
    GET(p->data.intent.failure_code, INT_MAX);
    GET(p->data.intent.root, LONG_MAX);
    GET(p->data.intent.state_before, INT_MAX);
    GET(p->data.intent.state_after, INT_MAX);
    GET(p->data.intent.delay_used_after, INT_MAX);
    if (!text(in, p->data.intent.context_sha256, sizeof p->data.intent.context_sha256)) return 0;
    GET(p->data.intent.intent_present, INT_MAX);
    if (!text(in, p->data.intent.intent_sha256, sizeof p->data.intent.intent_sha256)) return 0;
    GET(p->data.intent.intent_sha256_present, INT_MAX);
    GET(p->data.intent.failure_code_present, INT_MAX);
    GET(p->data.intent.intent.op, INT_MAX);
    GET(p->data.intent.intent.state, INT_MAX);
            break;
        case 4:
    GET(p->data.effect.family, INT_MAX);
    GET(p->data.effect.outcome, INT_MAX);
    GET(p->data.effect.root, LONG_MAX);
    GET(p->data.effect.activation_monstermoves, LONG_MAX);
    GET(p->data.effect.m_id, UINT_MAX);
            break;
        case 5:
    GET(p->data.termination.failure_code, INT_MAX);
    GET(p->data.termination.reason, INT_MAX);
    GET(p->data.termination.slot_f, INT_MAX);
    GET(p->data.termination.slot_w, INT_MAX);
    GET(p->data.termination.w_runtime, INT_MAX);
            break;
        default: return 0;
        }
    }
    for (i = 0; i < r->public_count; ++i) {
        struct chaos_next_use_public_record *p = &r->public_records[i];
    GET(p->next_use_public_v, INT_MAX);
    GET(p->family, INT_MAX);
    GET(p->phase, INT_MAX);
    GET(p->root, LONG_MAX);
    GET(p->notice_seq, LONG_MAX);
    GET(p->end_seq, LONG_MAX);
    }
    return 1;
}
static int load_snapshot(const char *name, struct chaos_next_use_snapshot *s) {
    int fd = open(name, O_RDONLY), ok; unsigned char extra;
    if (fd < 0) return 0;
    ok = chaos_next_use_snapshot_read(fd, s) && read(fd, &extra, 1) == 0;
    close(fd); return ok;
}
/* Compare every persistent gameplay value against the ACCEPTED replay prefix.
 * The actual saved OPEN journal anchor is separately checked against its exact
 * trusted prefix before this process starts; the validator has no writer.
 * Import uses that original anchor verbatim, not a fabricated NONE/COMPLETE.
 */
static int checkpoint_equal(const struct chaos_next_use_snapshot *s) {
    if (s->program_id != replay_runtime.program_id) { fprintf(stderr, "checkpoint program_id\n"); return 0; }
    if (s->phase != replay_runtime.phase) { fprintf(stderr, "checkpoint phase\n"); return 0; }
    if (s->slot_w != replay_runtime.slot_w) { fprintf(stderr, "checkpoint slot_w\n"); return 0; }
    if (s->slot_f != replay_runtime.slot_f) { fprintf(stderr, "checkpoint slot_f\n"); return 0; }
    if (s->w_runtime != replay_runtime.w_runtime) { fprintf(stderr, "checkpoint w_runtime\n"); return 0; }
    if (s->state != replay_runtime.state) { fprintf(stderr, "checkpoint state\n"); return 0; }
    if (s->delay_used != replay_runtime.delay_used) { fprintf(stderr, "checkpoint delay_used\n"); return 0; }
    if (s->callback_ordinal != replay_runtime.callback_ordinal) { fprintf(stderr, "checkpoint callback_ordinal\n"); return 0; }
    if (s->witnessed != replay_runtime.witnessed) { fprintf(stderr, "checkpoint witnessed\n"); return 0; }
    if (s->attention_claimed != replay_runtime.attention_claimed) { fprintf(stderr, "checkpoint attention_claimed\n"); return 0; }
    if (s->whistle_count != replay_runtime.whistle_count) { fprintf(stderr, "checkpoint whistle_count\n"); return 0; }
    if (s->fountain_count != replay_runtime.fountain_count) { fprintf(stderr, "checkpoint fountain_count\n"); return 0; }
    if (s->next_seq != replay_runtime.next_seq) { fprintf(stderr, "checkpoint next_seq\n"); return 0; }
    if (s->termination_emitted != replay_runtime.termination_emitted) { fprintf(stderr, "checkpoint termination_emitted\n"); return 0; }
    if (s->identity_unsafe != replay_runtime.identity_unsafe) { fprintf(stderr, "checkpoint identity_unsafe\n"); return 0; }
    if (s->callback_w != replay_runtime.callback_w) { fprintf(stderr, "checkpoint callback_w\n"); return 0; }
    if (s->callback_f != replay_runtime.callback_f) { fprintf(stderr, "checkpoint callback_f\n"); return 0; }
    if (s->last_root != replay_runtime.last_root) { fprintf(stderr, "checkpoint last_root\n"); return 0; }
    if (s->admission_move != replay_runtime.admission_move) { fprintf(stderr, "checkpoint admission_move\n"); return 0; }
    if (s->program_expiry != replay_runtime.program_expiry) { fprintf(stderr, "checkpoint program_expiry\n"); return 0; }
    if (s->delay_until != replay_runtime.delay_until) { fprintf(stderr, "checkpoint delay_until\n"); return 0; }
    if (s->variant != replay_runtime.variant) { fprintf(stderr, "checkpoint variant\n"); return 0; }
    if (s->origin_w_live != replay_runtime.origin_w_live) { fprintf(stderr, "checkpoint origin_w_live\n"); return 0; }
    if (s->origin_f_live != replay_runtime.origin_f_live) { fprintf(stderr, "checkpoint origin_f_live\n"); return 0; }
    if (s->armed_m_id != replay_runtime.armed_m_id) { fprintf(stderr, "checkpoint armed_m_id\n"); return 0; }
    if (s->replay_cursor != replay_runtime.replay_cursor) { fprintf(stderr, "checkpoint replay_cursor\n"); return 0; }
    if (s->origin_w != replay_runtime.origin_w) { fprintf(stderr, "checkpoint origin_w\n"); return 0; }
    if (s->origin_f != replay_runtime.origin_f) { fprintf(stderr, "checkpoint origin_f\n"); return 0; }
    if (s->origin_w_deadline != replay_runtime.origin_w_deadline) { fprintf(stderr, "checkpoint origin_w_deadline\n"); return 0; }
    if (s->origin_f_deadline != replay_runtime.origin_f_deadline) { fprintf(stderr, "checkpoint origin_f_deadline\n"); return 0; }
    if (s->run_token != replay_runtime.run_token) { fprintf(stderr, "checkpoint run_token\n"); return 0; }
    if (s->level_token != replay_runtime.level_token) { fprintf(stderr, "checkpoint level_token\n"); return 0; }
    if (s->activation_monstermoves != replay_runtime.activation_monstermoves) { fprintf(stderr, "checkpoint activation_monstermoves\n"); return 0; }
    if (s->armed_root != replay_runtime.armed_root) { fprintf(stderr, "checkpoint armed_root\n"); return 0; }
    if (s->source_length != replay_runtime.source_length) { fprintf(stderr, "checkpoint source_length\n"); return 0; }
    return !strcmp(s->source_sha256, replay_runtime.source_sha256)
        && !strcmp(s->binding_sha256, replay_runtime.binding_sha256)
        && !memcmp(s->source, replay_runtime.source, s->source_length)
        && !replay_runtime.pending_w_capture && !replay_runtime.f_inflight
        && !replay_runtime.defer_termination
        && s->journal_state == CHAOS_JOURNAL_OPEN && !s->capture_incomplete;
}
int main(int argc, char **argv) {
    struct chaos_next_use_snapshot initial, middle;
    struct chaos_next_use_replay_input record;
    struct runtime_state before, live_before;
    uint64_t count, n;
    FILE *in; char magic[6]; int checkpoints = 0, terminal;
    if (argc != 4 || !load_snapshot(argv[1], &initial)
        || !load_snapshot(argv[2], &middle)
        || !chaos_next_use_snapshot_import(&initial)) return 2;
    in = fopen(argv[3], "rb");
    if (!in || fread(magic, 1, 6, in) != 6 || memcmp(magic, "NUSP1", 6)
        || !number(in, &count, 4096) || !count) return 2;
    for (n = 0; n < count; ++n) {
        if (!record_read(in, &record)) return 2;
        /* Clock values are inputs, NOT an independent clock oracle. Python
         * binds them to retained native input-boundary clock intervals. */
        monstermoves = record.at_move;
        before = replay_runtime; live_before = live_runtime;
        if (chaos_next_use_replay_record(&record) != CHAOS_REPLAY_APPLIED) {
            int unchanged = !memcmp(&before, &replay_runtime, sizeof before)
                && !memcmp(&live_before, &live_runtime, sizeof live_before);
            const char *reason = "private_or_public_carrier_mismatch";
            if (record.cursor != before.replay_cursor + 1) reason = "cursor";
            else if (record.state != staged_runtime.state)
                reason = "source_dependent_next_state";
            else if (!replay_post_equal(&record.post, &staged_runtime))
                reason = "poststate";
            printf("{\"accepted\":%lu,\"cursor\":%lu,\"rejected_cursor\":%lu,"
                   "\"operation\":%d,\"reason\":\"%s\",\"unchanged\":%d}\n",
                   (unsigned long)n, before.replay_cursor, record.cursor,
                   record.operation, reason, unchanged);
            fclose(in); return unchanged ? 1 : 2;
        }
        if (record.cursor == middle.replay_cursor) {
            if (!checkpoint_equal(&middle)
                || !chaos_next_use_snapshot_import(&middle)) return 2;
            ++checkpoints;
        }
    }
    if (fgetc(in) != EOF) return 2;
    fclose(in);
    terminal = replay_runtime.phase == CHAOS_ATTEMPT_TERMINATED
        && replay_runtime.termination_emitted && !replay_runtime.pending_w_capture
        && !replay_runtime.f_inflight && replay_runtime.slot_w != CHAOS_SLOT_W_PENDING
        && replay_runtime.slot_f != CHAOS_SLOT_F_PENDING
        && replay_runtime.w_runtime != CHAOS_W_RUNTIME_ARMED;
    printf("{\"accepted\":%lu,\"cursor\":%lu,\"checkpoints\":%d,\"terminal\":%d}\n",
           (unsigned long)count, replay_runtime.replay_cursor, checkpoints, terminal);
    return terminal && checkpoints == 1 ? 0 : 2;
}

File: tests/chaos/next_use_semantic_preflight.py
"""Test-only value bridge to the actual C runtime, never native effect commands.

The native codec is host-specific, just like the retained native save. The
transition bridge uses bounded network-order integers/length-prefixed strings.
No journal text is compiled or evaluated. Independent bundle authentication is
required by the caller; self-rehashed negatives are NOT authentic recordings.
"""

import copy
import hashlib
import json
from pathlib import Path
import shutil
import struct
import subprocess

from chaos.next_use_journal import read_journal, JournalError, _TRANSITION, _POST

ROOT = Path(__file__).resolve().parents[2]
TRANSITION = [
    k
    for k in _TRANSITION.split()
    if k not in ("expected_token", "post", "private_records", "public_records")
]
POST = _POST.split()
PRIVATE = "next_use_private_v kind seq at_move program_id source_sha256".split()
INTENT = (
    "callback_ordinal trigger validation failure_code root state_before "
    "state_after delay_used_after context_sha256 intent_present intent_sha256 "
    "intent_sha256_present failure_code_present"
).split()
EFFECT = "family outcome root activation_monstermoves m_id".split()
TERMINATION = "failure_code reason slot_f slot_w w_runtime".split()
PUBLIC = "next_use_public_v family phase root notice_seq end_seq".split()
TOKEN = "root active remap consumed".split()
SNAP = (
    "snapshot_v program_id phase slot_w slot_f w_runtime state delay_used "
    "callback_ordinal admission_move program_expiry delay_until variant source_length "
    "origin_w_live origin_f_live origin_w origin_f origin_w_deadline origin_f_deadline "
    "armed_m_id replay_cursor run_token level_token activation_monstermoves armed_root "
    "witnessed attention_claimed whistle_count fountain_count next_seq termination_emitted "
    "identity_unsafe last_root callback_w callback_f run_token_high journal_state "
    "capture_incomplete journal_bytes"
).split()


def sha(raw):
    return hashlib.sha256(raw).hexdigest()


def native_snapshot(save, source, destination, prefix, exported):
    """Locate one exact source in a trusted save; C subsequently reads its codec."""
    raw = save.read_bytes()
    assert raw.count(source) == 1, "ambiguous native snapshot source"
    start = raw.index(source) - 355  # native v5: 40 int32 + three 65-byte hashes
    assert start >= 0
    codec = raw[start : start + 355 + len(source)]
    values = dict(zip(SNAP, struct.unpack("=40I", codec[:160])))
    values["run_token"] |= values.pop("run_token_high") << 32
    for index, key in enumerate(("source_sha256", "binding_sha256", "journal_sha256")):
        b = codec[160 + index * 65 : 225 + index * 65]
        assert b[-1] == 0
        values[key] = b.rstrip(b"\0").decode("ascii")
    assert values["snapshot_v"] == 5 and values["source_length"] == len(source)
    assert values["source_sha256"] == sha(source)
    for key in values.keys() & exported.keys():
        assert values[key] == exported[key], "native/export binding: " + key
    prefix_bytes = prefix.read_bytes()
    lines = prefix_bytes.splitlines(keepends=True)
    parsed = read_journal(prefix)
    # An actual terminal save includes the end row, which consumes no cursor.
    # Preserve its COMPLETE anchor verbatim; never normalize it to OPEN.
    complete = parsed["status"] == "structurally_complete"
    assert parsed["status"] in ("incomplete", "structurally_complete")
    assert values["replay_cursor"] == len(parsed["records"]) - (2 if complete else 1)
    assert values["journal_state"] == (2 if complete else 1)
    assert values["capture_incomplete"] == 0
    assert values["journal_bytes"] == sum(map(len, lines))
    assert values["journal_sha256"] == json.loads(lines[-1])["sha256"]
    destination.write_bytes(codec)
    return values


def build(directory):
    binary = directory / "semantic-preflight"
    if binary.exists():
        return binary
    source = ROOT / "tests/chaos/next_use_semantic_preflight.c"
    shutil.copy2(source, directory / source.name)
    shutil.copy2(__file__, directory / Path(__file__).name)
    command = [
        "cc",
        "-DCHAOS",
        "-ffunction-sections",
        "-fdata-sections",
        "-Wall",
        "-Wextra",
        "-Werror",
        "-Wno-misleading-indentation",
        "-isystem",
        str(ROOT / "include"),
        str(source),
    ]
    command += [
        str(ROOT / ("src/" + n + ".c"))
        for n in (
            "chaos_next_use",
            "chaos_next_use_admission",
            "chaos_protocol",
            "chaos_lua",
        )
    ]
    command += [
        "-Wl,--gc-sections",
        *subprocess.check_output(
            ["pkg-config", "--cflags", "--libs", "lua5.4"], text=True
        ).split(),
        "-lm",
        "-o",
        str(binary),
    ]
    (directory / "semantic-build-command.json").write_text(
        json.dumps(command, indent=2)
    )
    inputs = [
        source,
        ROOT / "src/chaos_next_use_runtime.c",
        ROOT / "tests/chaos/next_use_replay.c",
    ]
    inputs += [Path(s) for s in command if s.endswith(".c")]
    (directory / "semantic-build-inputs.json").write_text(
        json.dumps({str(p): sha(p.read_bytes()) for p in inputs}, indent=2)
    )
    result = subprocess.run(command, capture_output=True, text=True, timeout=45)
    (directory / "semantic-build.log").write_text(result.stdout + result.stderr)
    assert result.returncode == 0, result.stderr
    return binary


def wire(records, path):
    output = bytearray(b"NUSP1\0")

    def value(v):
        if isinstance(v, str):
            raw = v.encode("ascii")
            assert len(raw) <= 64
            output.extend(struct.pack(">Q", len(raw)))
            output.extend(raw)
        else:
            assert type(v) is int and 0 <= v <= 9223372036854775807
            output.extend(struct.pack(">Q", v))

    def fields(v, names):
        for k in names:
            value(v[k])

    transitions = [r["data"] for r in records if r["kind"] == "transition"]
    assert 0 < len(transitions) <= 4096
    value(len(transitions))
    for t in transitions:
        fields(t, TRANSITION)
        fields(t["post"], POST)
        fields(t["expected_token"], TOKEN)
        assert len(t["private_records"]) == t["private_count"] <= 4
        assert len(t["public_records"]) == t["public_count"] <= 1
        for r in t["private_records"]:
            fields(r, PRIVATE)
            fields(r["data"], {3: INTENT, 4: EFFECT, 5: TERMINATION}[r["kind"]])
            if r["kind"] == 3:
                fields(r["data"]["intent"], ["op", "state"])
        for r in t["public_records"]:
            fields(r, PUBLIC)
    path.write_bytes(output)


def run_validator(records, directory, label="authentic"):
    path = directory / (label + ".wire")
    wire(records, path)
    result = subprocess.run(
        [
            str(build(directory)),
            str(directory / "initial.codec"),
            str(directory / "middle.codec"),
            str(path),
        ],
        capture_output=True,
        text=True,
        timeout=10,
    )
    (directory / (label + ".log")).write_text(result.stdout + result.stderr)
    assert result.returncode in (0, 1), result.stdout + result.stderr
    report = json.loads(result.stdout)
    report["returncode"] = result.returncode
    return report


def validate_bundle(decoded, files, directory):
    source = files["source"].read_bytes()
    complete = files["record/next_use-journal.jsonl"].read_bytes()
    for key in (
        "prefix/next_use-journal.jsonl",
        "middle-prefix/next_use-journal.jsonl",
    ):
        assert complete.startswith(files[key].read_bytes()), (
            "checkpoint prefix binding: " + key
        )
    states = json.loads(files["states.json"].read_text())
    schedule = json.loads(files["schedule.json"].read_text())
    middle = [s for s, op in zip(states, schedule) if op["kind"] == "save"]
    assert len(middle) == 1
    initial = native_snapshot(
        files["save"],
        source,
        directory / "initial.codec",
        files["prefix/next_use-journal.jsonl"],
        json.loads(files["initial"].read_text()),
    )
    native_snapshot(
        next(p for k, p in files.items() if k.startswith("middle-save/")),
        source,
        directory / "middle.codec",
        files["middle-prefix/next_use-journal.jsonl"],
        middle[0],
    )
    header = decoded["records"][0]["data"]["snapshot"]
    # NONE is the original pre-writer header; OPEN is the acknowledged native
    # saved prefix. Compare gameplay values, but never rewrite either anchor.
    for k in initial:
        if k not in ("journal_state", "journal_bytes", "journal_sha256"):
            assert initial[k] == header[k], "header/native binding: " + k
    assert initial["replay_cursor"] == 0, (
        "this bounded pair requires its genuine header-only start"
    )
    # Independent native observations bound clocks at each input/session edge.
    previous = json.loads(files["initial"].read_text())
    ts = [r["data"] for r in decoded["records"] if r["kind"] == "transition"]
    for state in states:
        for t in ts[previous["replay_cursor"] : state["replay_cursor"]]:
            assert previous["monstermoves"] <= t["at_move"] <= state["monstermoves"]
        previous = state
    result = run_validator(decoded["records"], directory)
    assert result["returncode"] == 0, result
    (directory / "semantic-result.json").write_text(json.dumps(result, indent=2))
    return result


def rehash(records, path):
    previous, lines = "0" * 64, []
    for r in records:
        r["prev"] = previous
        payload = json.dumps(r, separators=(",", ":"), ensure_ascii=True).encode()
        previous = sha(payload)
        lines.append(
            b'{"payload":' + payload + b',"sha256":"' + previous.encode() + b'"}\n'
        )
    path.write_bytes(b"".join(lines))


def exercise_negatives(test, decoded, files, directory):
    reports = []
    for name in ("source_next_state", "cursor_duplicate", "poststate", "context"):
        records = copy.deepcopy(decoded["records"])
        transitions = [r["data"] for r in records if r["kind"] == "transition"]
        if name == "source_next_state":
            action = next(
                t for t in transitions if t["operation"] == 1 and t["family"] == 2
            )
            for t in transitions[action["cursor"] - 1 :]:
                t["state"] = 2
            d = action["private_records"][0]["data"]
            d["state_after"] = d["intent"]["state"] = 2
            d["intent_sha256"] = sha(
                json.dumps(
                    {"next_use_intent_v": 2, "op": "fountain_refresh", "state": 2},
                    sort_keys=True,
                    separators=(",", ":"),
                ).encode()
            )
        elif name == "cursor_duplicate":
            transitions[1]["cursor"] = 1
        elif name == "poststate":
            transitions[0]["post"]["expected_manifest_m_id"] = 1
        else:
            action = next(t for t in transitions if t["operation"] == 1)
            action["private_records"][0]["data"]["context_sha256"] = "0" * 64
        path = directory / ("validator-negative-" + name + ".jsonl")
        rehash(records, path)
        try:
            read_journal(path)
            strict = "passed"
        except JournalError as exc:
            strict = str(exc)
        test.assertEqual(strict == "passed", name in ("source_next_state", "poststate"))
        # Deliberately independent: no weakening of strict reader or trusted
        # digest gate to admit these fabricated validator-negative recordings.
        report = run_validator(records, directory, "validator-negative-" + name)
        test.assertEqual(report["returncode"], 1, report)
        test.assertEqual(report["unchanged"], 1, report)
        expected = {
            "source_next_state": (35, 36, "source_dependent_next_state"),
            "cursor_duplicate": (1, 1, "cursor"),
            "poststate": (0, 1, "poststate"),
            "context": (2, 3, "private_or_public_carrier_mismatch"),
        }[name]
        test.assertEqual(
            (report["cursor"], report["rejected_cursor"], report["reason"]), expected
        )
        reports.append(
            dict(
                name=name, strict_python=strict, native_playback_started=False, **report
            )
        )
    (directory / "validator-negatives.json").write_text(json.dumps(reports, indent=2))
